File: gitdot-server/src/handler/organization/list_organization_members.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
};

use gitdot_api::endpoint::list_organization_members as api;
use gitdot_core::dto::ListMembersRequest;

use crate::{
    app::{AppError, AppResponse, AppState},
    dto::IntoApi,
};

#[axum::debug_handler]
pub async fn list_organization_members(
    State(state): State<AppState>,
    Path(org_name): Path<String>,
    Query(params): Query<api::ListOrganizationMembersRequest>,
) -> Result<AppResponse<api::ListOrganizationMembersResponse>, AppError> {
    let request = ListMembersRequest::new(
        &org_name,
        params.role.as_deref(),
        params.cursor.as_deref(),
        params.limit,
    )?;
    state
        .org_service
        .list_members(request)
        .await
        .map_err(AppError::from)
        .map(|page| AppResponse::new(StatusCode::OK, page.into_api()))
}

File: gitdot-web/packages/gitdot-api/src/endpoint/organization/list-organization-members.ts
import { z } from "zod";
import { OrganizationMemberResource, page } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListOrganizationMembersRequest = z.object({
  role: z.string().optional(),
  cursor: z.string().optional(),
  limit: z.number().int().positive().optional(),
});
export type ListOrganizationMembersRequest = z.infer<
  typeof ListOrganizationMembersRequest
>;

export const ListOrganizationMembersResponse = page(OrganizationMemberResource);
export type ListOrganizationMembersResponse = z.infer<
  typeof ListOrganizationMembersResponse
>;

export const ListOrganizationMembers = {
  path: "/organization/{org_name}/members",
  method: "GET",
  request: ListOrganizationMembersRequest,
  response: ListOrganizationMembersResponse,
} as const satisfies Endpoint;
export type ListOrganizationMembers = typeof ListOrganizationMembers;
